# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/XlibDisplay.cpp
# Commit: c1bd0996764c
# Full Hash: c1bd0996764c49174c9169fe5550905ce5dcef88
# Author: Jed Davis <jld@mozilla.com>
# Date: 2021-07-06 21:42:42
# Regressor Bug: 1635451
# File Overlap Count: 1
# Description:
#   Bug 1635451 - Allow GLX to work in headless content processes. r=jgilbert
#   
#   This follows what we're already doing for EGL: a refcounted object
#   which can own the X connection, where we hold a weak reference from
#   the library object so that multiple contexts opportunistically share
# ==============================================================================

diff -r 2974b1ba2beb -r c1bd0996764c gfx/thebes/XlibDisplay.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/gfx/thebes/XlibDisplay.cpp	Tue Jul 06 07:42:42 2021 +0000
@@ -0,0 +1,40 @@
+/* -*- Mode: C++; tab-width: 20; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "XlibDisplay.h"
+
+#include "mozilla/Assertions.h"
+
+namespace mozilla::gfx {
+
+XlibDisplay::XlibDisplay(Display* aDisplay, bool aOwned)
+    : mDisplay(aDisplay), mOwned(aOwned) {
+  MOZ_ASSERT(mDisplay);
+}
+
+XlibDisplay::~XlibDisplay() {
+  if (mOwned) {
+    XCloseDisplay(mDisplay);
+  }
+}
+
+/* static */
+std::shared_ptr<XlibDisplay> XlibDisplay::Borrow(Display* aDisplay) {
+  if (!aDisplay) {
+    return nullptr;
+  }
+  return std::shared_ptr<XlibDisplay>(new XlibDisplay(aDisplay, false));
+}
+
+/* static */
+std::shared_ptr<XlibDisplay> XlibDisplay::Open(const char* aDisplayName) {
+  Display* disp = XOpenDisplay(aDisplayName);
+  if (!disp) {
+    return nullptr;
+  }
+  return std::shared_ptr<XlibDisplay>(new XlibDisplay(disp, true));
+}
+
+}  // namespace mozilla::gfx